### ISSUE TITLE
Explain optionals?

### ISSUE THREAD

**Issue by gratimax** [SpongePowered/SpongeDocs#355]
Continuing from [a previous issue](https://github.com/SpongePowered/SpongeDocs/issues/123#issuecomment-76640936).

I've noticed that many plugin developers do not know how to use `java.util.Optional` correctly, and it appears throughout the API. There already exist [guides](https://www.voxxed.com/blog/2015/05/why-even-use-java-8-optional/) online that explain Optional, but I think that having an introduction in the documentation would assist those trying to work with it, and also provide a Sponge-curated resource to link to.

Some topics that might be covered:
- What `Optional<T>` means: a value that may or may not be there, meaning that _your plugin_ has to handle both cases somehow.
- Why you should call `Optional#get()` very carefully: throwing hard-to-debug errors, essentially lifting NullPointerException into a NoSuchElementException
- Using `Optional#orElse` and `Optional#orElseThrow`: orElse for a default value, orElseThrow to throw a _descriptive exception_ that you should handle later
- The standard pattern of:

    ```java
    if (opt.isPresent()) {
      T value = opt.get();
    } else {
       // handle failure
    }
```

Some more, potentially advanced topics:
- Using `Optional#ifPresent`: Only do something if it exists, no errors otherwise
- Using `Optional#flatMap` to chain optionals: useful especially with Data API, you can use this to chain an optional value with methods that return additional methods, i.e. `opt.flatMap(value -> value.getData(SomeData.class))`
- Constructing your own optionals: using `Optional.empty()` for empty, `Optional.of` for present, and `Optional.ofNullable` to lift a null into an Optional

**Comment by gabizou** [comment]
Don't forget to mention the following:

```java
    @Listener
    public void demo(InteractEntityEvent entityEvent) {
        entityEvent.getCause().first(Player.class).ifPresent(player -> {
            final ItemStack itemStack = entityEvent.getGame().getRegistry().createItemBuilder()
                .itemType(ItemTypes.DIAMOND_SWORD)
                .quantity(1)
                .build();
            List<ItemEnchantment> enchantments = new ArrayList<>();
            enchantments.add(new ItemEnchantment(Enchantments.SHARPNESS, 100));
            enchantments.add(new ItemEnchantment(Enchantments.FIRE_ASPECT, 10));
            itemStack.offer(Keys.ITEM_ENCHANTMENTS, enchantments);
            
            player.setItemInHand(itemStack);
        });
    }

```
Is equivalent to:
```java
    @Listener
    public void altdemo(InteractEntityEvent entityEvent) {
        if (entityEvent.getCause().any(Player.class)) {
            final Player player = entityEvent.getCause().first(Player.class).get();
            final ItemStack itemStack = entityEvent.getGame().getRegistry().createItemBuilder()
                .itemType(ItemTypes.DIAMOND_SWORD)
                .quantity(1)
                .build();
            List<ItemEnchantment> enchantments = new ArrayList<>();
            enchantments.add(new ItemEnchantment(Enchantments.SHARPNESS, 100));
            enchantments.add(new ItemEnchantment(Enchantments.FIRE_ASPECT, 10));
            itemStack.offer(Keys.ITEM_ENCHANTMENTS, enchantments);

            player.setItemInHand(itemStack);
        }
    }
```

The advantage is that it's all wrapped as a consumer item eliminating the need for an if/else statement.

**Comment by Tzky** [comment]
This is a very good idea as some developers don't seem to get the idea behind using ``optional``\s at all. To them this is just added work without any benefits, while ``optional``\s are very useful.

**Comment by Inscrutable** [comment]
There appears to be general consensus amongst active Docs staff that this would be a useful addition. Options would need its' own page(s), but we can add links to it from many other pages to get it the attention it deserves.

**Comment by boformer** [comment]
I agree.

**Comment by boformer** [comment]
I agree.

**Comment by Mumfrey** [comment]
I made an attempt: [link](https://gist.github.com/Mumfrey/eaee54bb6dbb3a4242a6)

**Comment by Zidane** [comment]
@Mumfrey your explanations are wonderful as always.

**Comment by Tzky** [comment]
Honestly this is by far better than anything i could've come up with. We should get this into the docs asap.
What about the images? Any license issues or can we use them aswell?

**Comment by Mumfrey** [comment]
You don't have to use the images if you don't want to, I just put them in to make the content less boring.

**Comment by Zidane** [comment]
@Mumfrey 

Images make everything! Can't be removed :P.

**Comment by gratimax** [comment]
@gabizou That isn't a result of optionals, that's the cause-specific API. That would belong in another document explaining causes, because this one is only about optionals.

@Mumfrey That looks awesome, explains optionals really well. Thanks

**Comment by saladoc** [comment]
@gratimax Sure, gabizou used cause as an example. However `Cause#first(...)` returns an `Optional`, so it is perfectly legitimate to ask that we document `Optional#ifPresent(Consumer)` as well.

**Comment by gratimax** [comment]
@saladoc Absolutely, I agree. It's already under the todo list in "advanced topics". :stuck_out_tongue:

**Comment by ryantheleach** [comment]
Those images are great, can't we get like an id on each of the text bubbles, so translations can be provided in a caption?

**Comment by Tzky** [comment]
What about adding the text bubbles below the image? 
Ex:
[Image here]
A: "blabla"
B: "blub"
A: "Whoa!"

This would leave the images untouched (i like them including the text) but would provide a possibility to translate them.

**Comment by boformer** [comment]
I think we can assume that the readers understand simple English. Also, the images are not important for understanding.

**Comment by Tzky** [comment]
*Also, the article just kind of ends abruptly because I would anticipate someone else finishing it up with the more advanced topics mentioned above. I just don't feel I could write those up particularly well so I leave them as an excercise for someone else.*

Any thoughts on the scope of our Optional docs? Should we include the advanced topics or just link elsewhere? Currently i've just copied and converted Mumfreys to reST.

**Comment by Inscrutable** [comment]
The page is already quite large, so breaking it into multiple pages would give a good excuse to leave the more complicated use-cases for a separate doc. Linking to external resources to explain the intricacies of native optionals would also be a good idea in general.

**Comment by saladoc** [comment]
We may split it up in two pages - one is provided by mumfrey, the excellent explanation why `Optional`s are a thing and why we prefer them over returning `null`. The second page could be the code examples with a short explanation, detailing the possibilities of the `Optional` class rather than the rationale behind it.

**Comment by Tzky** [comment]
This was resolved in #359
